x-pack/plugins/lens/public/heatmap_visualization/chart_component.tsx

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { FC, useEffect, useState } from 'react';
+import { i18n } from '@kbn/i18n';
+import {
+  Chart,
+  ElementClickListener,
+  Heatmap,
+  HeatmapBrushEvent,
+  HeatmapElementEvent,
+  HeatmapSpec,
+  ScaleType,
+  Settings,
+} from '@elastic/charts';
+import { VisualizationContainer } from '../visualization_container';
+import { HeatmapRenderProps } from './types';
+import './index.scss';
+import { LensBrushEvent, LensFilterEvent } from '../types';
+import { desanitizeFilterContext } from '../utils';
+import { search } from '../../../../../src/plugins/data/public';
+
+declare global {
+  interface Window {
+    /**
+     * Flag used to enable debugState on elastic charts
+     */
+    _echDebugStateFlag?: boolean;
+  }
+}
+
+export const HeatmapComponent: FC<HeatmapRenderProps> = ({
+  data,
+  args,
+  timeZone,
+  formatFactory,
+  onClickValue,
+  onSelectRange,
+}) => {
+  const isDarkTheme = false;
+
+  const table = Object.values(data.tables)[0];
+
+  const chartData = table.rows;
+
+  const xAxisColumnIndex = table.columns.findIndex((v) => v.id === args.xAccessor);
+  const yAxisColumnIndex = table.columns.findIndex((v) => v.id === args.yAccessor);
+
+  const xAxisColumn = table.columns[xAxisColumnIndex];
+  const yAxisColumn = table.columns[yAxisColumnIndex];
+  const valueColumn = table.columns.find((v) => v.id === args.valueAccessor);
+
+  if (!xAxisColumn || !valueColumn) {
+    // Chart is not ready
+    return null;
+  }
+
+  if (!yAxisColumn) {
+    chartData.forEach((row) => {
+      row.unifiedY = i18n.translate('xpack.lens.heatmap.emptyYLabel', {
+        defaultMessage: '(empty)',
+      });
+    });
+  }
+
+  const xAxisMeta = xAxisColumn.meta;
+  const isTimeBasedSwimLane = xAxisMeta.type === 'date';
+  const xScaleType = isTimeBasedSwimLane ? ScaleType.Time : ScaleType.Ordinal;
+
+  const xValuesFormatter = formatFactory(xAxisMeta.params);
+  const valueFormatter = formatFactory(valueColumn.meta.params);
+
+  const xDomain = (() => {
+    if (!isTimeBasedSwimLane) return null;
+    const dateInterval = search.aggs.getDateHistogramMetaDataByDatatableColumn(xAxisColumn)
+      ?.interval;
+    if (!dateInterval) return null;
+    const intervalDuration = search.aggs.parseInterval(dateInterval);
+    if (!intervalDuration) return null;
+
+    return {
+      min: data.dateRange?.fromDate.getTime(),
+      max: data.dateRange?.toDate.getTime(),
+      minInterval: intervalDuration.as('milliseconds'),
+    };
+  })();
+
+  // @ts-ignore
```

Comment:
I had to add it because of the union type of the click handler 
![image](https://user-images.githubusercontent.com/5236598/118953263-f2688d00-b95c-11eb-94b1-41db410ee062.png)